### PR TITLE
Removing lex tuples, cf. #1916

### PR DESCRIPTION
**Pull request by aseemr**

### CODE REVIEW

File: src/typechecker/FStar.TypeChecker.TcTerm.fs [hunk 1]
```diff
@@ -3270,7 +3333,7 @@ and build_let_rec_env _top_level env lbs : list<letbinding> * env_t * guard_t =
             then g_acc, t
             else (let env0 = Env.push_univ_vars env0 univ_vars in
                   let t, _, g = tc_check_tot_or_gtot_term ({env0 with check_uvars=true}) t (fst <| U.type_u()) "" in
-                  Env.conj_guard g_acc (g |> Rel.resolve_implicits env |> Rel.discharge_guard env), norm env0 t) in
```
Comment (aseemr) [inline src/typechecker/FStar.TypeChecker.TcTerm.fs:3273]: @nikswamy , this is the `norm` call I removed, please take a look.

File: src/smtencoding/FStar.SMTEncoding.Term.fs [hunk 2]
```diff
@@ -1013,18 +1013,10 @@ and mkPrelude z3options =
...
                        |> List.map (declToSmt z3options) |> String.concat "\n" in
-   let lex_ordering = "\n(define-fun is-Prims.LexCons ((t Term)) Bool \n\
-                                   (is-LexCons t))\n\
-                       (declare-fun Prims.lex_t () Term)\n\
-                       (assert (forall ((t1 Term) (t2 Term) (x1 Term) (x2 Term) (y1 Term) (y2 Term))\n\
-                                    (iff (Valid (Prims.precedes Prims.lex_t Prims.lex_t (LexCons t1 x1 x2) (LexCons t2 y1 y2)))\n\
-                                         (or (Valid (Prims.precedes t1 t2 x1 y1))\n\
-                                             (and (= x1 y1)\n\
-                                                  (Valid (Prims.precedes Prims.lex_t Prims.lex_t x2 y2)))))))\n\
+   let lex_ordering = "\n(declare-fun Prims.lex_t () Term)\n\
```
Comment (aseemr) [inline src/smtencoding/FStar.SMTEncoding.Term.fs:1019]: Also, note that the smtencoding still has a `lex_t` type, though it is not exposed to the user. Basically, we emit queries for `precedes t1 t2 x y`, and switch to `precedes lex_t lex_t x y` as before.

Comment (aseemr) [reply src/smtencoding/FStar.SMTEncoding.Term.fs:1019]: @nikswamy

File: src/typechecker/FStar.TypeChecker.TcTerm.fs (hunk 1 above)
Comment (nikswamy) [reply src/typechecker/FStar.TypeChecker.TcTerm.fs:3273]: Looks fine. I'm curious what prompted you to remove it.

File: src/smtencoding/FStar.SMTEncoding.Term.fs (hunk 2 above)
Comment (nikswamy) [reply src/smtencoding/FStar.SMTEncoding.Term.fs:1019]: I think this lex_t is just a dummy type constant at this point. Maybe we can just remove it? And use Prims.unit for this purpose instead?

**Comment by nikswamy** [comment]
Hi Aseem, this is really awesome to have this near completion with so much cleanup, and importantly, closing the unsoundness bug #1916.

Some suggestions to put the cherry on the top of this and merge it:

It would be great to have a description here of the main elements of our solution, just for future reference. In particular, I would find it very helpful to have a brief guide on the following points:
  - Explain a bit the removal of the lex tuples altogether, in favor of desugaring the decreases clause
  - Where does the desugaring of the decreases clause with lex tuples take place? There's new parsing rules for it etc. right? This would be particularly useful to know as we try to extend decreases clauses further with user-provided relations
  - What were the examples that needed to be adapted, type annotations added and the new warning on unequal types etc.
  - CHANGES.md summarizing this and pointing back to this PR for more details

Thanks!

**Comment by aseemr** [comment]
## Overview

This PR fixes the treatment of lexicographic tuples in the typechecker.

In the current implementation of the lex tuples, F* primitive library `prims.fst` defines a list-like inductive type `lex_t`. The lexicographic tuples in the programs are desugared into values of these types, which are then ordered in the `precedes` relation by an axiom in the smt encoding. As the issue https://github.com/FStarLang/FStar/issues/1916 shows, exposing `lex_t` directly can lead to unsoundness.

In this PR, the main idea is to remove the inductive type `lex_t` and desugar the decreases clauses directly in the typechecker. Specifically, given two decreases clauses `%[x1; x2; x3]` and `%[y1; y2; y3]`, whereas earlier we would have asked SMT to prove:

```
LexCons x1 (LexCons x2 (LexCons x3 LexTop))

<<

LexCons y1 (LexCons y2 (LexCons y3 LexTop))
```

by relying on an axiom about `LexCons` etc. in the smt encoding, we now ask SMT to directly prove:

```
x1 << y1 \/ (x1 == y1 /\ x2 << y2) \/ (x1 == y1 /\ x2 == y2 /\ x3 << y3)
```

This allows us to cleanup the primitive library by removing the `lex_t` inductive type (and hence `LexCons`, `LexTop` etc.). This also means that the user programs can no longer build the infinite, unsound chains as in the issue https://github.com/FStarLang/FStar/issues/1916. All the programmer can do now is to write these "abstract" decreases clauses.


## Typechecker code changes

### Syntax

The decreases clauses in the previous syntax were represented as a `term`, both in `FStar.Parser.AST` and `FStar.Syntax.Syntax`. At the time of desugaring, we were taking the lexicographic lists that the programmer wrote, and creating a term such as `LexCons x1 (LexCons ...)`, and that's what the rest of the typechecker manipulated.

In the new syntax, we represent decreases clauses as lists both in `FStar.Parser.AST` and `FStar.Syntax.AST`. In the parser AST, this PR adds a new node `LexList of (list term)`. The parser on the `%[` token, parses the list of terms, and creates a `LexList` which is then embedded in the `Decreases of term` node, making it something like `Decreases (LexList [x1; x2; x3])`.

Similarly, in `FStar.Syntax.Syntax`, the `DECREASES` flag maintains a `list Syntax.term`.

### Typechecker

These lists are desugared, typechecked, etc. in the expected manner.

When we are building the type environment for the body of the recursive function, at that time we fold over these lists to create the precedes term from them. This happens in `FStar.Typechecker.TcTerm.fs` in the `guard_letrecs` function. This code essentially takes two lexicographic lists and builds the term `x1 << y1 \/ (x1 === y1 /\ ...)` shown above. We use `Prims.eq3` for the equality, and this formula is then used to refine the type of the recursive function as before.

This change also allows us to several small, but useful, cleanups. E.g. we used to typecheck the `lex_t` inductive type specially, that whole function is gone. There were several `if type = lex_t` kind of special casing that we could get rid of.

### SMT Encoding

With this PR, the axioms that we were adding w.r.t. `LexCons` etc. are gone. We also made it more solid in that the equality terms `x1 == y1` in the precedes formula, are now `Prims.eq3`, and hence typed, whereas earlier the axiom about `LexCons`  used to ignore the types altogether.


## Changes in the client code

In most cases, the client code remains unchanged since the usual way of writing the decreases clauses remains the same (i.e. `%[...]`).

However, since we made the equality in the precedes formula typed, there are some examples that need some tweaks. For example, consider:

```
module Test

let g (n:nat) : nat = n
let h (n:nat) : int = n

let rec foo (n:nat) (m:nat) : Tot nat (decreases %[g n; g m]) =
  if m = 0 then 0
  else bar n (m - 1)

and bar (n:nat) (m:nat) : Tot nat (decreases %[h n; h m]) = 0
```

This is a roundabout way of writing the lexicographic order of the arguments using the indirection of `g` and `h` functions. Crucially, the return type of `g` is `nat`, whereas for `h` it is `int`.

The recursive call `bar` in the body of `foo` requires us to prove this:

```
h n << g n \/ (h n === gn /\ h m << g m)
```

If we expand the `===` to also write types, it is actually:

```
h n << g n \/ (Prims.eq3 #int #nat (h n) (g n) /\ h m == g m)
```

And now the proof is stuck, since `Prims.eq3` also enforces that the two type arguments are equal, and here clearly `int` and `nat` are not equal.

In such cases, F* now gives a warning. In this particular example, when we try to verify it, F* fails, but emits a warning as follows:

```
(Warning) SMT may not be able to prove the types of Test.g n at <input>(6,51-6,54) (Prims.nat) and Test.h n at <input>(10,47-10,50) (Prims.int) to be equal, if the proof fails, try annotating these with the same type
```

as a hint that the proof may require an annotation. In this case, we can change the type of `g` or `h` to be the same, or even better, just annotate the decreases clause, as follows:

```
let rec foo (n:nat) (m:nat) : Tot nat (decreases %[(g n) <: int; g m]) =
```

By annotating `(g n) <: int`, the arguments to `Prims.eq3` will now both be `int`, and the proof works. Here are a couple of examples in the F* test suite that required such changes: https://github.com/FStarLang/FStar/pull/2218/commits/0baf2277cd1e2c83ba71c4bc9659f1a84837a33a.

**Pull request merged**